### PR TITLE
feat(cache/oss): add support for oss

### PR DESCRIPTION
**Pull request by PsiACE**
- [x] add basic support for oss, pass local test
- [x] polish impl & docs
- [ ] add test workflow

Fixes: #2043

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,25 +23,28 @@ sccache is also available as a [GitHub Actions](https://github.com/marketplace/a
...
-  * [S3](docs/S3.md)
-  * [R2](docs/S3.md#R2)
-  * [Redis](docs/Redis.md)
-  * [Memcached](docs/Memcached.md)
-  * [Google Cloud Storage](docs/Gcs.md)
-  * [Azure](docs/Azure.md)
-  * [GitHub Actions](docs/GHA.md)
-  * [WebDAV (Ccache/Bazel/Gradle compatible)](docs/Webdav.md)
+- [sccache - Shared Compilation Cache](#sccache---shared-compilation-cache)
+- [Table of Contents (ToC)](#table-of-contents-toc)
```
Comment (Xuanwo) [inline README.md:27]: It's better not to introduce unrelated changes. Welcome to submit them as seperate PR.

File: docs/OSS.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# OSS
+
+If you want to use OSS storage for the sccache cache, you need to set the `SCCACHE_OSS_BUCKET` environment variable to the name of the OSS bucket to use.
+
+You **must** configure the region using the `SCCACHE_ENDPOINT` environment variable.
```
Comment (Xuanwo) [inline docs/OSS.md:5]: Are we sharing the same `SCCACHE_ENDPOINT` environment? Seems we are using `SCCACHE_OSS_ENDPOINT`

File: src/cache/oss.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+impl OSSCache {
+    pub fn build(
+        bucket: &str,
+        key_prefix: &str,
+        endpoint: Option<&str>,
+    ) -> Result<Operator> {
+        let mut builder = Oss::default();
+        builder.bucket(bucket);
+        builder.root(key_prefix);
+
```
Comment (Xuanwo) [inline src/cache/oss.rs:30]: It's better for us to support setting `access_key_id` here.

Comment (PsiACE) [reply src/cache/oss.rs:31]: Regarding the S3 configuration, I have included the `no_credentials` option. When this option is set to false, it necessitates loading two corresponding environment variables. If set to true, it should be skipped.

I believe there is no need to provide the `access_key_id` in the build function.

**Review by Xuanwo — APPROVED** [review]
Thanks a lot!

**Comment by Xuanwo** [comment]
cc @sylvestre, would you like to take a review? OSS doesn't provide something like minio or [localstack](https://github.com/localstack/localstack), so we can't test it in CI.

File: docs/OSS.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# OSS
+
+If you want to use OSS storage for the sccache cache, you need to set the `SCCACHE_OSS_BUCKET` environment variable to the name of the OSS bucket to use.
+
+You **must** specify the endpoint URL using the `SCCACHE_OSS_ENDPOINT` environment variable. More details about OSS endpoints [here](https://www.alibabacloud.com/help/en/oss/user-guide/regions-and-endpoints).
```
Comment (sylvestre) [inline docs/OSS.md:5]:
```suggestion
You **must** specify the endpoint URL using the `SCCACHE_OSS_ENDPOINT` environment variable. More details about [OSS endpoints](https://www.alibabacloud.com/help/en/oss/user-guide/regions-and-endpoints).
```

Using "Here" isn't great for screen reader (accessbility)

File: docs/OSS.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# OSS
+
+If you want to use OSS storage for the sccache cache, you need to set the `SCCACHE_OSS_BUCKET` environment variable to the name of the OSS bucket to use.
```
Comment (sylvestre) [inline docs/OSS.md:3]:
```suggestion
If you want to use _Object Storage Service_ (aka OSS) for the sccache cache, you need to set the `SCCACHE_OSS_BUCKET` environment variable to the name of the OSS bucket to use.
```

File: src/cache/oss.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+use opendal::layers::LoggingLayer;
+use opendal::services::Oss;
+use opendal::Operator;
+
+use crate::errors::*;
+
+pub struct OSSCache;
+
+impl OSSCache {
```
Comment (sylvestre) [inline src/cache/oss.rs:21]:
```suggestion
// Implement the Object Storage Service for Alibaba cloud
impl OSSCache {
```

**Comment by sylvestre** [comment]
Some minor nits :)

**Comment by sylvestre** [comment]
thanks for your work!
